plutus-benchmark/src/Plutus/Benchmark/Queens.hs

```diff
@@ -0,0 +1,553 @@
+earliestInconsistency CSP{rel=rel} [] = Nothing
+earliestInconsistency CSP{rel=rel} (a:as) =
+        case filter (not . rel a) (reverse as) of
+          []    -> Nothing
+          (b:_) -> Just (level a, level b)
+
+{-# INLINABLE labelInconsistencies #-}
+labelInconsistencies :: CSP -> Transform State (State,Maybe (Var,Var))
+labelInconsistencies csp = mapTree f
+    where f s = (s,earliestInconsistency csp s)
+
+{-# INLINABLE btsolver0 #-}
+btsolver0 :: CSP -> [State]
+btsolver0 csp =
+  (filter (complete csp) . leaves . (mapTree fst) . prune ((/= Nothing) . snd)
+                                            . (labelInconsistencies csp) .  mkTree) csp
+
+-----------------------------------------------
+-- Figure 6. Conflict-directed solving of CSPs.
+-----------------------------------------------
+
+data ConflictSet = Known [Var] | Unknown
+instance Eq ConflictSet where
+    Known v == Known w = v == w
+    Unknown == Unknown = True
+    _ == _ = False
+
+{-# INLINABLE knownConflict #-}
+knownConflict :: ConflictSet -> Bool
+knownConflict (Known (a:as)) = True
+knownConflict _              = False
+
+{-# INLINABLE knownSolution #-}
+knownSolution :: ConflictSet -> Bool
+knownSolution (Known []) = True
+knownSolution _          = False
+
+{-# INLINABLE checkComplete #-}
+checkComplete :: CSP -> State -> ConflictSet
+checkComplete csp s = if complete csp s then Known [] else Unknown
+
+type Labeler = CSP -> Transform State (State, ConflictSet)
+
+{-# INLINABLE search #-}
+search :: Labeler -> CSP -> [State]
+search labeler csp =
+  (map fst . filter (knownSolution . snd) . leaves . prune (knownConflict . snd) . labeler csp . mkTree) csp
+
+{-# INLINABLE bt #-}
+bt :: Labeler
+bt csp = mapTree f
+      where f s = (s,
+                   case earliestInconsistency csp s of
+                     Nothing    -> checkComplete csp s
+                     Just (a,b) -> Known [a,b])
+
+{-# INLINABLE btsolver #-}
+btsolver :: CSP -> [State]
+btsolver = search bt
+
+-------------------------------------
+-- Figure 7. Randomization heuristic.
+-------------------------------------
+
+{-# INLINABLE hrandom #-}
+hrandom :: Integer -> Transform a a
+hrandom seed (Node a cs) = Node a (randomList seed' (zipWith hrandom (randoms (length cs) seed') cs))
+  where seed' = random seed
+
+{-# INLINABLE btr #-}
+btr :: Integer -> Labeler
+btr seed csp = bt csp . hrandom seed
+
+---------------------------------------------
+-- Support for random numbers (not in paper).
+---------------------------------------------
+
+{-# INLINABLE random2 #-}
+random2 :: Integer -> Integer
+random2 n = if test > 0 then test else test + 2147483647
+  where test = 16807 * lo - 2836 * hi
+        hi   = n `div` 127773
+        lo   = n `rem` 127773
+
+{-# INLINABLE randoms #-}
+randoms :: Integer -> Integer -> [Integer]
+randoms k = iterateN k random2
+
+{-# INLINABLE random #-}
+random :: Integer -> Integer
+random n = (a * n + c) -- mod m
```

Comment:
What's up with `mod`?

Reply:
No idea!  That's a comment from the original file: https://github.com/ghc/nofib/blob/master/spectral/constraints/Main.hs